### ISSUE TITLE
Windows 10 1903 Window Drag Issue

### ISSUE THREAD

**Issue by HEIC-to-JPEG-Dev** [sourcechord/FluentWPF#42]
Hi, I just upgraded to Windows 10 1903 and noticed that using the titlebar to drag a window has a very high latency, lagging behind the cursor movements by seconds.

I'm using the AcrylicWindow with ExtendViewIntoTitleBar="True"

Thanks for any updates.

**Comment by Andrei-Lapu** [comment]
Experiencing the same issue on my 1903 build, could not yet figure out a solution.

**Comment by vivaladan** [comment]
I've only just discovered FluentWPF but have hit this issue right away on 1903 with a super simple new app.

**Comment by sourcechord** [comment]
Hi, @DHitchenUK 
Sorry for replying so late.
Thanks for the information.
I recognized your reporting issue.
I'm inspecting the behavior of FluentWPF in Windows10 19H1, and trying to fix it. 
Please wait for a while.

**Comment by DHitchenUK** [comment]
@sourcechord, I just wanted to say thanks for the work you put in, it's very much appreciated.

**Comment by dsafa** [comment]
If you turn off transparency effects in the settings, it works like before. Of course this turns off blur for all windows apps however.

**Comment by sourcechord** [comment]
Hi, @DHitchenUK 
Maybe, it is same as riverar/sample-win32-acrylicblur#2

FluentWPF uses undocumented Win32API(SetWindowCompositionAttribute method).

The reported behavior is caused by the change in the behavior of that API.
It seems that SetWindowCompositionAttribute API is broken in Windows10 1903.

Dragging problem occurs Windows10's emoji toolbar, too.
(that is shown by Win+period key)
And, these problem is reported in Feedback Hub.
https://aka.ms/AA4pzyf
https://aka.ms/AA4fmz7

Please upvote it 🙏

**Comment by ashelleyPurdue** [comment]
I'm still getting this issue even after updating to FluentWPF 0.7.0.

**Comment by ghost1372** [comment]
This bug seems to be fixed in version 1903 (18362.267)

**Comment by 0xCA** [comment]
> 
> 
> This bug seems to be fixed in version 1903 (18362.267)

18362.295, still here

**Comment by JVimes** [comment]
Is this supposed to be open? 637ea17 did not fix it for me (Windows 1903, OS build 18995.1)

**Comment by Armin2208** [comment]
In 1909, or a new cumulative update, the problems are again present.

**Comment by sourcechord** [comment]
Hi, @Armin2208

Thanks for the information.
I fixed it in FluentWPF v0.8.0-alpha004
Please try it.

**Comment by Armin2208** [comment]
I've tried it and it works. Nice, thank you!

**Comment by sourcechord** [comment]
I'm glad to hear that.
I'll close this issue.
If you have any problems, please reopen this issue.

**Comment by NickAcPT** [comment]
I know that this issue has been closed already, but is there a way to keep the real acrylic blur and not have these issues?

For example, I've seen DevExpress avoid this issue by temporarily disabling the transparency when moving the window, showing the Fallback colour.

**Comment by tricky-mind** [comment]
@sourcechord , I actually know how to recreate the acrylic effect of uwp using interop(not the setWindowComposition() method but using the direct x)

But have the same issue. But only on resizing and not on dragging

[Preview](https://firebasestorage.googleapis.com/v0/b/highq-b5355.appspot.com/o/2020-08-04-12-36-34.gif?alt=media&token=9b98ff25-3978-42b6-86f2-d8037e3843f4) 

This is totally different from your code.

**Comment by tricky-mind** [comment]
> @sourcechord , I actually know how to recreate the acrylic effect of uwp using interop(not the setWindowComposition() method but using the direct x)
> 
> But have the same issue. But only on resizing and not on dragging
> 
> [Preview](https://firebasestorage.googleapis.com/v0/b/highq-b5355.appspot.com/o/2020-08-04-12-36-34.gif?alt=media&token=9b98ff25-3978-42b6-86f2-d8037e3843f4)
> 
> This is totally different from your code.

[Acrylic Effect WPF](https://github.com/microsoft/Windows.UI.Composition-Win32-Samples/tree/master/dotnet/WPF/AcrylicEffect)

Just have a look @ this.

**Comment by ghost1372** [comment]
> > @sourcechord , I actually know how to recreate the acrylic effect of uwp using interop(not the setWindowComposition() method but using the direct x)
> > But have the same issue. But only on resizing and not on dragging
> > [Preview](https://firebasestorage.googleapis.com/v0/b/highq-b5355.appspot.com/o/2020-08-04-12-36-34.gif?alt=media&token=9b98ff25-3978-42b6-86f2-d8037e3843f4)
> > This is totally different from your code.
> 
> [Acrylic Effect WPF](https://github.com/microsoft/Windows.UI.Composition-Win32-Samples/tree/master/dotnet/WPF/AcrylicEffect)
> 
> Just have a look @ this.

This requires nuget packages

**Comment by tricky-mind** [comment]
> > > @sourcechord , I actually know how to recreate the acrylic effect of uwp using interop(not the setWindowComposition() method but using the direct x)
> > > But have the same issue. But only on resizing and not on dragging
> > > [Preview](https://firebasestorage.googleapis.com/v0/b/highq-b5355.appspot.com/o/2020-08-04-12-36-34.gif?alt=media&token=9b98ff25-3978-42b6-86f2-d8037e3843f4)
> > > This is totally different from your code.
> > 
> > 
> > [Acrylic Effect WPF](https://github.com/microsoft/Windows.UI.Composition-Win32-Samples/tree/master/dotnet/WPF/AcrylicEffect)
> > Just have a look @ this.
> 
> This requires nuget packages

Yes of course you need nuget packages, i believe they do it by capturing the screen below the window visual using direct x and blurring the image by using direct2d effects.

**Comment by VS-ux** [comment]
@tricky-mind 
I used SetWindowCompositionAttribute for my winforms app and since I was using insider builds, the dragging was fixed. (Dragging lag caused by mouse polling rate greater than monitor refresh rate)
However, resizing still posed issues. For some reason when my app is 32 bit I don't have any issues.

**Comment by sourcechord** [comment]
Hi there.
I restored acrylic blur effect in [FluentWPF 0.10.0-alpha001](https://github.com/sourcechord/FluentWPF/releases/tag/0.10.0-alpha001): this is pre-release packages.
Windows10 still has laggy behavior with ACRYLICBLURBEHIND. But  we can avoid laggy behavior by disabling acrylic blur when resizing or moving the window.

This is it 🎉 
![AcrylicBlur](https://user-images.githubusercontent.com/14825436/124390823-3b677b80-dd28-11eb-994d-e36c001cf428.gif)

I'm confirming this way in several environments. If there is no issues, I'll release it as FluentWPF 0.10.0.
Please wait for a while.

**Comment by NotARobot6969** [comment]
For me, it is opaque until dragged, which is the opposite of the intended effect.

https://user-images.githubusercontent.com/62841684/124396688-fe3cc280-dd02-11eb-8929-1ed072ec9dad.mp4

**Comment by NotARobot6969** [comment]
It is also worth mentioning that dragging the window on the desktop does not lag at all.

**Comment by sourcechord** [comment]
Hi @NotARobot6969

This is the intended behavior.
There are two types of blur effect;
* ACCENT_ENABLE_ACRYLICBLURBEHIND: Hard blur that is same as UWP's acrylic effect
* ACCENT_ENABLE_BLURBEHIND: Soft blur 

ACCENT_ENABLE_ACRYLICBLURBEHIND effect causes laggy behavior during window is moving or resizing. So that, FluentWPF applies ACCENT_ENABLE_ACRYLICBLURBEHIND effect only when the window is stabled, and applies ACCENT_ENABLE_BLURBEHIND effect during window is  moving or resizing.

**Comment by NotARobot6969** [comment]
@sourcechord But then there is no blur for most of the time, at least for me.

**Comment by NotARobot6969** [comment]
To clarify: For me, the laggy effect is applying when I move the window, and no effect is being applied when it's stationary.

**Comment by sourcechord** [comment]
According to the uploaded movie, the background of the window doesn't seems to have enough texture. It is difficult to recognize the blur effect.
Could you show the window on a more complex background images?

**Comment by NotARobot6969** [comment]
To be clear, it's lagging when it is being dragged. @sourcechord 
https://user-images.githubusercontent.com/62841684/124490002-837ab280-dda9-11eb-8314-147b7ffd798a.mp4

**Comment by sourcechord** [comment]
Please tell me your environment.
Your window has rounded border. Are you using windows11 insider preview?
I've not tested FluentWPF 0.10.0-alpha001 in windows11 yet 😓

**Comment by NotARobot6969** [comment]
Indeed! I forgot to mention it.

**Comment by NotARobot6969** [comment]
@sourcechord I forgot to tell you this too; the less windows behind the acrylic program, the less lag there is.

Desktop with no windows behind (smooth):
https://user-images.githubusercontent.com/62841684/124514339-b8e8c580-ddd4-11eb-873b-fc2f42053f53.mp4

Many windows behind (laggy):
https://user-images.githubusercontent.com/62841684/124514438-ecc3eb00-ddd4-11eb-98df-03fc8d108521.mp4

**Comment by Simnico99** [comment]
Seems to be a Windows 11 related issues as I am experiencing the same thing on Windows 11 Version 22000.51:

![18ab2e6be93ef481aac39f1269afd949](https://user-images.githubusercontent.com/80013536/124651725-483fb880-de69-11eb-929d-97ab090a73c9.gif)

**Comment by fhnb16** [comment]
@Simnico99 

Sorry for offtop, how you make rounded window borders with FluentWPF in Windows 11?

**Comment by alexmercerind** [comment]
> Seems to be a Windows 11 related issues as I am experiencing the same thing on Windows 11 Version 22000.51

Latest build of Windows 11 22000.100, acrylic has stopped working. 😭

**Comment by sourcechord** [comment]
I'm checking the window behavior with FluentWPF in Windows11 environment. But, currently there is no way to apply aclylic effect with SetWindowCompositionAttribute API.
So that I turnd off acrylic effect in FluentWPF 0.10.0-alpha002(only in Windows11 env).

It's a tough situation; however I'll continue to survey to provide 'Fluent Design' in windows11 environment.

**Comment by Extrimis** [comment]
@sourcechord i am working on a workaround using direct composition, a sample win32 demo is available here : [Acrylic Effect using DWM and Direct Composition](https://github.com/Extrimis/Win32-Acrylic-Effect). I am trying to build a wpf version, its a bit hard.

**Comment by alexmercerind** [comment]
@Extrimis I also tried to use similar approach but with XAML islands instead.
And those widgets always overlap the existing widgets. Can't figure out.

**Comment by Extrimis** [comment]
> @Extrimis I also tried to use similar approach but with XAML islands instead.
> And those widgets always overlap the existing widgets. Can't figure out.

Create Separate window and attach both windows together with SetWindowPos(), instead of Resizing Both Windows, Resize the Windows.UI.Composition.Visual so there wont be any flickering on resize. The acrylic window should be WS_EX_TRANSPARENT,WS_EX_LAYERED

**Comment by Extrimis** [comment]
> > @Extrimis I also tried to use similar approach but with XAML islands instead.
> > And those widgets always overlap the existing widgets. Can't figure out.
> 
> Create Separate window and attach both windows together with SetWindowPos(), instead of Resizing Both Windows, Resize the Windows.UI.Composition.Visual so there wont be any flickering on resize. The acrylic window should be WS_EX_TRANSPARENT,WS_EX_LAYERED

I am looking forward to mica than acrylic.

**Comment by sourcechord** [comment]
Hi, @Extrimis 
Thanks for the information, it was very interesting. But it seems hard to implement in WPF.

> I am looking forward to mica than acrylic.

I think so too. I'd like to introduce mica materials in FluentWPF, but haven't found a way to achieve it.

BTW FluentWPF currently has some problems in Windows11 that need to be fixed. So that, I created another GitHub issue(#130) to track it.
Please let me know about Windows 11 problems and acrylic effects in above issue.

**Comment by Extrimis** [comment]
Have you tried windows.ui.composition interop method for acrylic, that one won't lag but it has airspace issues that must be solved, it's the real acrylic

**Comment by sourcechord** [comment]
No, I heven't tried yet. Because it seems that Composition API doesn't support interop with WPF window.
I read following article, but there are only `CompositionHostControl` way, and no explanation for interop with WPF window.
https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/using-the-visual-layer-with-wpf

**Comment by sourcechord** [comment]
OIC 😃 
I tried [WPF AcrylicEffect Sample](https://github.com/microsoft/Windows.UI.Composition-Win32-Samples/tree/master/dotnet/WPF/AcrylicEffect), and recognized that can draw to window directly. But it has airspace issue as you said.
If we use Windows.UI.Composition, we have to create separete window for avoiding airspace issue. 

It is difficult to implement, and causes the increase of complexity.
So that, I'll release FluentWPF 0.10.0 with `SetWindowCompositionAttribute` way, however continue to investigate real acrylic effect.

**Comment by Extrimis** [comment]
Ok

**Comment by gen3vra** [comment]
has anyone else observed the behavior that in visual studio 2022 the drag is fine in debug mode?

in my application which makes use of SetWindowCompositionAttribute() I've observed the following strange behavior:
In visual studio 2022 debug mode Acrylic blur windows seem to function as originally intended.
Additionally, attaching the visual studio 2022 debugger to the built application makes the faulty build work for as long as the debugger is attached.
This behavior persisted in both Net 6 and Net core 3.1, and Visual Studio 2019 did not “fix” the application as the video shows 2022 did.
Video of behavior:
https://streamable.com/recib9
(ShareX basically gives up recording the laggy drag movement bug)
Can anyone let me know if they’re also seeing this?

![](https://i.imgur.com/7Ybgouf.png)
